File: src/main/scala/org/glavo/llvm/ir/SequentialType.scala
package org.glavo.llvm.ir

import org.glavo.llvm.internal.Unsigned
import org.glavo.llvm.internal.ir.SequentialTypeImpl

abstract class SequentialType private[llvm](handle: Long) extends CompositeType(handle) {
  val numElements: Long@Unsigned = SequentialTypeImpl.getNumElements(handle)

  val elementType: Type = Type(SequentialTypeImpl.getElementType(handle))
}

File: internal/src/main/c++/ir/SequentialType.cpp
//
// Created by glavo on 18-3-1.
//

#include <org_glavo_llvm_internal_ir_SequentialTypeImpl.h>

#include <llvm/IR/Type.h>
#include <llvm/IR/DerivedTypes.h>

#include <llvm4s/util.h>

using namespace llvm4s;
using namespace llvm;

jlong Java_org_glavo_llvm_internal_ir_SequentialTypeImpl_getNumElements(JNIEnv *, jclass, jlong handle) {
    return get<SequentialType>(handle)->getNumElements();
}

jlong Java_org_glavo_llvm_internal_ir_SequentialTypeImpl_getElementType(JNIEnv *, jclass, jlong handle) {
    return handleOf(get<SequentialType>(handle)->getElementType());
}